security/advancedtls/provider_test.go

```diff
@@ -0,0 +1,213 @@
+				t.Errorf("copyFileContents(test.certFileBefore, certTmpPath): %v", err)
+			}
+			err = copyFileContents(test.keyFileBefore, keyTmpPath.Name())
+			if err != nil {
+				t.Errorf("copyFileContents(test.keyFileBefore, keyTmpPath): %v", err)
+			}
+			identityPemFileProvider, err := NewIdentityPemFileProvider(identityPemFileProviderOptions)
+			if err != nil {
+				t.Errorf("NewIdentityPemFileProvider(identityPemFileProviderOptions) failed: %v", err)
+			}
+			gotKM, err := identityPemFileProvider.KeyMaterial(context.Background())
+			if !cmp.Equal(*gotKM, test.wantKmBefore, cmp.AllowUnexported(big.Int{})) {
+				t.Errorf("provider.KeyMaterial() = %+v, want %+v", *gotKM, test.wantKmBefore)
+			}
+			// ------------------------Stage 1------------------------------------
+			err = copyFileContents(test.certFileAfter, certTmpPath.Name())
+			if err != nil {
+				t.Errorf("copyFileContents(test.certFileAfter, certTmpPath): %v", err)
+			}
+			err = copyFileContents(test.keyFileAfter, keyTmpPath.Name())
+			if err != nil {
+				t.Errorf("copyFileContents(test.keyFileAfter, keyTmpPath): %v", err)
+			}
+			time.Sleep(5 * time.Second)
+			gotKM, err = identityPemFileProvider.KeyMaterial(context.Background())
+			if !cmp.Equal(*gotKM, test.wantKmAfter, cmp.AllowUnexported(big.Int{})) {
+				t.Errorf("provider.KeyMaterial() = %+v, want %+v", *gotKM, test.wantKmAfter)
+			}
+			identityPemFileProvider.Close()
+		})
+	}
+}
+
+// The RootPemFileProvider is tested in different stages.
+// At stage 0, we copy the first trustFile to the temp file that are watched by the goroutine.
+// At stage 1, we copy the second trustFile to the temp file and verify the credential files are updated.
+func (s) TestRootPemFileProvider(t *testing.T) {
+	cs := &certStore{}
+	err := cs.loadCerts()
+	if err != nil {
+		t.Errorf("cs.loadCerts() failed: %v", err)
+	}
+	// Create temp files that are used to hold root credentials.
+	trustTmpPath, err := ioutil.TempFile(os.TempDir(), "pre-")
+	if err != nil {
+		t.Errorf("ioutil.TempFile(os.TempDir(), pre-) failed: %v", err)
+	}
+	defer os.Remove(trustTmpPath.Name())
+	tests := []struct {
+		desc            string
+		trustFileBefore string
+		wantKmBefore    certprovider.KeyMaterial
+		trustFileAfter  string
+		wantKmAfter     certprovider.KeyMaterial
+	}{
+		{
+			desc:            "Test root provider on the client side",
+			trustFileBefore: testdata.Path("client_trust_cert_1.pem"),
+			wantKmBefore:    certprovider.KeyMaterial{Roots: cs.clientTrust1},
+			trustFileAfter:  testdata.Path("client_trust_cert_2.pem"),
+			wantKmAfter:     certprovider.KeyMaterial{Roots: cs.clientTrust2},
+		},
+		{
+			desc:            "Test root provider on the server side",
+			trustFileBefore: testdata.Path("server_trust_cert_1.pem"),
+			wantKmBefore:    certprovider.KeyMaterial{Roots: cs.serverTrust1},
+			trustFileAfter:  testdata.Path("server_trust_cert_2.pem"),
+			wantKmAfter:     certprovider.KeyMaterial{Roots: cs.serverTrust2},
+		},
```

Comment:
We can also create a case to check when the updated file contents are empty, the update shouldn't be applied.
Also, what if the initial file is empty? Does the call to `rootPemFileProvider.KeyMaterial` give us an empty key material? Can we also add test for that?

Reply:
I have updated the testing logic and it's commented above the tests. 
Stage 0 covers the case of "`rootPemFileProvider.KeyMaterial` gives us an empty key material when the initial file is empty". 
Stage 3 covers the case of "when the updated file contents are empty, the update shouldn't be applied".

Reply:
Yeah sound good. Nice change!